pkg/controller/base_controller.go

```diff
@@ -89,10 +100,30 @@ func (c *controller) Run(ctx context.Context) error {
 	// Waitgroup for starting controller goroutines.
 	var wg sync.WaitGroup
 
+	defer func() {
+		c.logger.Info("Waiting for all workers to shutdown")
+
+		c.workQueue.ShutDown()
+
+		// wait for all the go routines
+		wg.Wait()
+
+		c.logger.Info("All workers are down")
+
+	}()
+
 	wg.Add(c.numWorker)
 	for i := 0; i < c.numWorker; i++ {
 		go func() {
-			wait.Until(c.runWorker, c.syncPeriod, ctx.Done())
+			wait.Until(c.runWorker, c.reconcilePeriod, ctx.Done())
```

Comment:
runWorker is infinite loop.. we don't need wait.Until right?

Reply:
we need to use wait.Until for runWorker. As of now runWorker returns false if queue is marked for shutdown but it help us to handle panic and re-start the runWorker.